### PR TITLE
Make non-corepack yarn behave more like corepack yarn when packing git dependencies

### PR DESCRIPTION
**Pull request by charlessuh**
**What's the problem this PR addresses?**

If a git dependency does not have a `packageManager` field in its `package.json` (nor any `.yarnrc` or `.yarnrc.yml` files), corepack yarn will launch a new corepack yarn process which will read the package.json, see it does not have a `packageManager`, and pack using v1, while non-corepack yarn will launch a new yarn process which is whatever the currently running version of yarn is (which is at least v2) to pack.

This can lead to checksum mismatches (which throws due to `checksumBehavior: throw` being the default) as the version of yarn used to pack the git dependency is different depending on how yarn was initially installed (via corepack or via npm.)

...

**How did you fix it?**
Ensure yarn v1 is always used to pack a repository without a `packageManager` field.

...

**Checklist**

- [x] I have read the [Contributing Guide](https://yarnpkg.com/advanced/contributing).

- [x] I have set the packages that need to be released for my changes to be effective.

- [x] I will check that all automated PR checks pass before the PR gets reviewed.

### CODE REVIEW

File: packages/yarnpkg-core/sources/scriptUtils.ts [hunk 1]
```diff
@@ -230,10 +234,14 @@ export async function prepareExternalProject(cwd: PortablePath, outputPath: Port
               ? [`workspace`, workspace]
               : [];
 
-            // Makes sure that we'll be using Yarn 1.x
-            const version = await execUtils.pipevp(`yarn`, [`set`, `version`, `classic`, `--only-if-needed`], {cwd, env, stdin, stdout, stderr, end: execUtils.EndStrategy.ErrorCode});
```
Comment (merceyz) [inline packages/yarnpkg-core/sources/scriptUtils.ts:234]: The fix would be to make sure this command is executed using the current version of Yarn, we can't assume which version Corepack is set to use.

File: packages/yarnpkg-core/sources/scriptUtils.ts [hunk 2]
```diff
@@ -218,7 +222,7 @@ export async function prepareExternalProject(cwd: PortablePath, outputPath: Port
         effectivePackageManager = packageManagerSelection.packageManager;
       } else {
         stdout.write(`No package manager configuration detected; defaulting to Yarn\n\n`);
-        effectivePackageManager = PackageManager.Yarn2;
+        effectivePackageManager = PackageManager.Yarn1;
```
Comment (merceyz) [inline packages/yarnpkg-core/sources/scriptUtils.ts:225]: It should default to berry, not classic.

File: packages/yarnpkg-core/sources/scriptUtils.ts [hunk 3]
```diff
@@ -100,6 +101,9 @@ export async function detectPackageManager(location: PortablePath): Promise<Pack
   if (xfs.existsSync(ppath.join(location, `pnpm-lock.yaml` as PortablePath)))
     return {packageManager: PackageManager.Pnpm, reason: `found pnpm's "pnpm-lock.yaml" lockfile`};
 
+  if (xfs.existsSync(ppath.join(location, `.yarnrc.yml` as PortablePath)))
+    return {packageManager: PackageManager.Yarn2, reason: `found ".yarnrc.yml"`};
```
Comment (merceyz) [inline packages/yarnpkg-core/sources/scriptUtils.ts:105]: The existence of `.yarnrc.yml` doesn't mean the project uses berry.

File: packages/yarnpkg-core/sources/Cache.ts [hunk 4]
```diff
@@ -12,7 +12,7 @@ import * as miscUtils                                                   from './
 import * as structUtils                                                 from './structUtils';
 import {LocatorHash, Locator}                                           from './types';
 
-const CACHE_VERSION = 8;
+const CACHE_VERSION = 9;
```
Comment (merceyz) [inline packages/yarnpkg-core/sources/Cache.ts:15]: Updating the global cache version for this isn't something we should do, instead we should add a cache version to git dependencies so we only need to regenerate those.
Something like what we do for patch dependencies https://github.com/yarnpkg/berry/blob/7ae458b8165ad53a8ef9db0060cbb6de73305768/packages/plugin-patch/sources/PatchResolver.ts#L9

**Review by merceyz — CHANGES_REQUESTED** [review]

**Comment by arcanis** [comment]
> while non-corepack yarn will launch a new yarn process which is whatever the currently running version of yarn is (which is at least v2) to pack.

It'll check the lockfile and, if it seems Yarn 1, it'll first call `yarn set version classic` so that the followup pack runs on v1. That seems to be the expected behavior, right?

**Pull request closed without merging**